File: lib/quichex/h3.ex
defmodule Quichex.H3 do
  @moduledoc """
  Low-level HTTP/3 bindings built on top of `Quichex.Native`.
  """

  import Bitwise

  alias Quichex.Native

  @type h3_config :: reference()
  @type h3_connection :: reference()
  @type connection :: reference()
  @type stream_id :: non_neg_integer()
  @type header :: {binary(), binary()}
  @type h3_event ::
          {:headers, [header()], boolean()}
          | :data
          | :finished
          | {:reset, non_neg_integer()}
          | {:goaway, non_neg_integer()}
          | :priority_update

  @spec config_new() :: h3_config()
  def config_new do
    unwrap_ok!("h3_config_new", Native.h3_config_new())
  end

  @spec config_set_max_field_section_size(h3_config(), non_neg_integer()) :: h3_config()
  def config_set_max_field_section_size(config, value) do
    unwrap_ok!("h3_config_set_max_field_section_size", Native.h3_config_set_max_field_section_size(config, value), config)
  end

  @spec config_set_qpack_max_table_capacity(h3_config(), non_neg_integer()) :: h3_config()
  def config_set_qpack_max_table_capacity(config, value) do
    unwrap_ok!("h3_config_set_qpack_max_table_capacity", Native.h3_config_set_qpack_max_table_capacity(config, value), config)
  end

  @spec config_set_qpack_blocked_streams(h3_config(), non_neg_integer()) :: h3_config()
  def config_set_qpack_blocked_streams(config, value) do
    unwrap_ok!("h3_config_set_qpack_blocked_streams", Native.h3_config_set_qpack_blocked_streams(config, value), config)
  end

  @spec config_enable_extended_connect(h3_config(), boolean()) :: h3_config()
  def config_enable_extended_connect(config, enabled) do
    unwrap_ok!("h3_config_enable_extended_connect", Native.h3_config_enable_extended_connect(config, enabled), config)
  end

  @spec config_set_additional_settings(h3_config(), [{non_neg_integer(), non_neg_integer()}]) ::
          h3_config()
  def config_set_additional_settings(config, settings) do
    unwrap_ok!("h3_config_set_additional_settings", Native.h3_config_set_additional_settings(config, settings), config)
  end

  @spec conn_new_with_transport(connection(), h3_config()) :: h3_connection()
  def conn_new_with_transport(conn, h3_config) do
    unwrap_ok!("h3_conn_new_with_transport", Native.h3_conn_new_with_transport(conn, h3_config))
  end

  @spec conn_poll(connection(), h3_connection()) :: {stream_id(), h3_event()} | :done
  def conn_poll(conn, h3_conn) do
    case Native.h3_conn_poll(conn, h3_conn) do
      {:ok, {stream_id, {:h3_headers, headers, more}}} ->
        {stream_id, {:headers, headers, more}}

      {:ok, {stream_id, event}} ->
        {stream_id, event}

      {:error, "done"} -> :done
      {:error, reason} -> raise Quichex.Native.ConfigError, message: reason, reason: reason
    end
  end

  @spec send_request(connection(), h3_connection(), [header()], boolean()) :: stream_id()
  def send_request(conn, h3_conn, headers, fin) do
    unwrap_ok!("h3_send_request", Native.h3_send_request(conn, h3_conn, headers, fin))
  end

  @spec send_response(connection(), h3_connection(), stream_id(), [header()], boolean()) :: :ok
  def send_response(conn, h3_conn, stream_id, headers, fin) do
    unwrap_ok!("h3_send_response", Native.h3_send_response(conn, h3_conn, stream_id, headers, fin))
    :ok
  end

  @spec send_body(connection(), h3_connection(), stream_id(), binary(), boolean()) :: :ok
  def send_body(conn, h3_conn, stream_id, body, fin) do
    unwrap_ok!("h3_send_body", Native.h3_send_body(conn, h3_conn, stream_id, body, fin))
    :ok
  end

  @spec recv_body(connection(), h3_connection(), stream_id(), non_neg_integer()) ::
          binary() | :done
  def recv_body(conn, h3_conn, stream_id, max_bytes) do
    case Native.h3_recv_body(conn, h3_conn, stream_id, max_bytes) do
      {:ok, data} -> data
      {:error, "done"} -> :done
      {:error, reason} -> raise Quichex.Native.ConfigError, message: reason, reason: reason
    end
  end

  @spec send_goaway(connection(), h3_connection(), stream_id()) :: :ok
  def send_goaway(conn, h3_conn, stream_id) do
    unwrap_ok!("h3_send_goaway", Native.h3_send_goaway(conn, h3_conn, stream_id))
    :ok
  end

  @spec extended_connect_enabled_by_peer(h3_connection()) :: boolean()
  def extended_connect_enabled_by_peer(h3_conn) do
    unwrap_ok!("h3_extended_connect_enabled_by_peer", Native.h3_extended_connect_enabled_by_peer(h3_conn))
  end

  @spec dgram_enabled_by_peer(connection(), h3_connection()) :: boolean()
  def dgram_enabled_by_peer(conn, h3_conn) do
    unwrap_ok!("h3_dgram_enabled_by_peer", Native.h3_dgram_enabled_by_peer(conn, h3_conn))
  end

  @doc """
  Send an HTTP/3 datagram with the given `flow_id`.

  For WebTransport, the datagram `flow_id` is the CONNECT stream id.
  """
  @spec send_datagram(connection(), non_neg_integer(), binary()) :: :ok
  def send_datagram(conn, flow_id, payload) when is_integer(flow_id) and flow_id >= 0 do
    encoded = IO.iodata_to_binary([encode_varint(flow_id), payload])
    unwrap_ok!("connection_dgram_send", Native.connection_dgram_send(conn, encoded))
    :ok
  end

  @doc """
  Receive an HTTP/3 datagram and decode its `flow_id`.

  Returns `{:ok, flow_id, payload}` or `:done` when no datagram is available.
  """
  @spec recv_datagram(connection(), non_neg_integer()) ::
          {:ok, non_neg_integer(), binary()} | :done
  def recv_datagram(conn, max_len) when is_integer(max_len) and max_len > 0 do
    case Native.connection_dgram_recv(conn, max_len) do
      {:ok, data} ->
        case decode_varint(data) do
          {:ok, flow_id, rest} -> {:ok, flow_id, rest}
          {:error, reason} -> raise Quichex.Native.ConfigError, message: reason, reason: reason
        end

      {:error, "done"} ->
        :done

      {:error, reason} ->
        raise Quichex.Native.ConfigError, message: reason, reason: reason
    end
  end

  defp unwrap_ok!(function, result, fallback \\ nil) do
    case result do
      :ok -> fallback
      {:ok, value} -> value
      {:error, reason} ->
        raise Quichex.Native.ConfigError,
          message: "#{function} failed: #{reason}",
          reason: reason,
          function: function

      other ->
        raise Quichex.Native.ConfigError,
          message: "#{function} failed: #{inspect(other)}",
          reason: inspect(other),
          function: function
    end
  end

  defp encode_varint(value) when value < 0x40 do
    <<0b00::2, value::6>>
  end

  defp encode_varint(value) when value < 0x4000 do
    <<0b01::2, value::14>>
  end

  defp encode_varint(value) when value < 0x4000_0000 do
    <<0b10::2, value::30>>
  end

  defp encode_varint(value) when value < 0x4000_0000_0000_0000 do
    <<0b11::2, value::62>>
  end

  defp encode_varint(value) do
    raise Quichex.Native.ConfigError,
      message: "flow_id out of range: #{value}",
      reason: "flow_id out of range"
  end

  defp decode_varint(<<>>), do: {:error, "empty datagram"}

  defp decode_varint(<<first, rest::binary>>) do
    length = first >>> 6
    case length do
      0 -> {:ok, first &&& 0x3F, rest}
      1 -> decode_varint_n(first, rest, 2)
      2 -> decode_varint_n(first, rest, 4)
      3 -> decode_varint_n(first, rest, 8)
    end
  end

  defp decode_varint_n(_first, rest, len) when byte_size(rest) + 1 < len do
    {:error, "incomplete varint"}
  end

  defp decode_varint_n(first, rest, 2) do
    <<b1, tail::binary>> = rest
    value = ((first &&& 0x3F) <<< 8) ||| b1
    {:ok, value, tail}
  end

  defp decode_varint_n(first, rest, 4) do
    <<b1, b2, b3, tail::binary>> = rest
    value = ((first &&& 0x3F) <<< 24) ||| (b1 <<< 16) ||| (b2 <<< 8) ||| b3
    {:ok, value, tail}
  end

  defp decode_varint_n(first, rest, 8) do
    <<b1, b2, b3, b4, b5, b6, b7, tail::binary>> = rest
    value =
      ((first &&& 0x3F) <<< 56) ||| (b1 <<< 48) ||| (b2 <<< 40) ||| (b3 <<< 32) |||
        (b4 <<< 24) ||| (b5 <<< 16) ||| (b6 <<< 8) ||| b7

    {:ok, value, tail}
  end
end

File: native/quichex_nif/src/h3.rs
use crate::resources::{ConnectionResource, H3ConfigResource, H3ConnectionResource};
use quiche::h3::NameValue;
use rustler::{Binary, Encoder, Env, OwnedBinary, ResourceArc, Term};

rustler::atoms! {
    h3_headers,
    data,
    finished,
    reset,
    goaway,
    priority_update,
}

enum H3Event {
    Headers {
        headers: Vec<(Vec<u8>, Vec<u8>)>,
        more_frames: bool,
    },
    Data,
    Finished,
    Reset { error_code: u64 },
    GoAway { id: u64 },
    PriorityUpdate,
}

impl Encoder for H3Event {
    fn encode<'a>(&self, env: Env<'a>) -> Term<'a> {
        match self {
            H3Event::Headers {
                headers: header_list,
                more_frames,
            } => {
                let headers: Vec<(Binary, Binary)> = header_list
                    .iter()
                    .map(|(name, value)| (to_binary(env, name), to_binary(env, value)))
                    .collect();
                (h3_headers(), headers, *more_frames).encode(env)
            }
            H3Event::Data => data().encode(env),
            H3Event::Finished => finished().encode(env),
            H3Event::Reset { error_code } => (reset(), *error_code).encode(env),
            H3Event::GoAway { id } => (goaway(), *id).encode(env),
            H3Event::PriorityUpdate => priority_update().encode(env),
        }
    }
}

fn to_binary<'a>(env: Env<'a>, bytes: &[u8]) -> Binary<'a> {
    let mut owned =
        OwnedBinary::new(bytes.len()).expect("Failed to allocate binary for header");
    owned.as_mut_slice().copy_from_slice(bytes);
    Binary::from_owned(owned, env)
}

fn build_headers(headers: Vec<(Binary, Binary)>) -> Vec<quiche::h3::Header> {
    headers
        .into_iter()
        .map(|(name, value)| quiche::h3::Header::new(name.as_slice(), value.as_slice()))
        .collect()
}

fn h3_error_to_string(err: quiche::h3::Error) -> String {
    format!("{:?}", err)
}

#[rustler::nif]
pub fn h3_config_new() -> Result<ResourceArc<H3ConfigResource>, String> {
    quiche::h3::Config::new()
        .map(|config| {
            ResourceArc::new(H3ConfigResource {
                inner: std::sync::Arc::new(std::sync::Mutex::new(config)),
            })
        })
        .map_err(|e| format!("Failed to create H3 config: {:?}", e))
}

#[rustler::nif]
pub fn h3_config_set_max_field_section_size(
    config: ResourceArc<H3ConfigResource>,
    v: u64,
) -> Result<(), String> {
    let mut cfg = config
        .inner
        .lock()
        .map_err(|e| format!("Lock error: {}", e))?;
    cfg.set_max_field_section_size(v);
    Ok(())
}

#[rustler::nif]
pub fn h3_config_set_qpack_max_table_capacity(
    config: ResourceArc<H3ConfigResource>,
    v: u64,
) -> Result<(), String> {
    let mut cfg = config
        .inner
        .lock()
        .map_err(|e| format!("Lock error: {}", e))?;
    cfg.set_qpack_max_table_capacity(v);
    Ok(())
}

#[rustler::nif]
pub fn h3_config_set_qpack_blocked_streams(
    config: ResourceArc<H3ConfigResource>,
    v: u64,
) -> Result<(), String> {
    let mut cfg = config
        .inner
        .lock()
        .map_err(|e| format!("Lock error: {}", e))?;
    cfg.set_qpack_blocked_streams(v);
    Ok(())
}

#[rustler::nif]
pub fn h3_config_enable_extended_connect(
    config: ResourceArc<H3ConfigResource>,
    enabled: bool,
) -> Result<(), String> {
    let mut cfg = config
        .inner
        .lock()
        .map_err(|e| format!("Lock error: {}", e))?;
    cfg.enable_extended_connect(enabled);
    Ok(())
}

#[rustler::nif]
pub fn h3_config_set_additional_settings(
    config: ResourceArc<H3ConfigResource>,
    settings: Vec<(u64, u64)>,
) -> Result<(), String> {
    let mut cfg = config
        .inner
        .lock()
        .map_err(|e| format!("Lock error: {}", e))?;
    cfg.set_additional_settings(settings)
        .map_err(|e| format!("Failed to set additional settings: {:?}", e))
}

#[rustler::nif]
pub fn h3_conn_new_with_transport(
    conn: ResourceArc<ConnectionResource>,
    h3_config: ResourceArc<H3ConfigResource>,
) -> Result<ResourceArc<H3ConnectionResource>, String> {
    let mut quic_conn = conn
        .inner
        .lock()
        .map_err(|e| format!("Lock error: {}", e))?;
    let cfg = h3_config
        .inner
        .lock()
        .map_err(|e| format!("Lock error: {}", e))?;

    quiche::h3::Connection::with_transport(&mut *quic_conn, &*cfg)
        .map(|h3_conn| {
            ResourceArc::new(H3ConnectionResource {
                inner: std::sync::Arc::new(std::sync::Mutex::new(h3_conn)),
            })
        })
        .map_err(|e| format!("Failed to create H3 connection: {:?}", e))
}

#[rustler::nif]
pub fn h3_conn_poll(
    conn: ResourceArc<ConnectionResource>,
    h3_conn: ResourceArc<H3ConnectionResource>,
) -> Result<(u64, H3Event), String> {
    let mut quic_conn = conn
        .inner
        .lock()
        .map_err(|e| format!("Lock error: {}", e))?;
    let mut h3_conn = h3_conn
        .inner
        .lock()
        .map_err(|e| format!("Lock error: {}", e))?;

    match h3_conn.poll(&mut *quic_conn) {
        Ok((stream_id, event)) => {
            let event = match event {
                quiche::h3::Event::Headers { list, more_frames } => H3Event::Headers {
                    headers: list
                        .iter()
                        .map(|h| (h.name().to_vec(), h.value().to_vec()))
                        .collect(),
                    more_frames,
                },
                quiche::h3::Event::Data => H3Event::Data,
                quiche::h3::Event::Finished => H3Event::Finished,
                quiche::h3::Event::Reset(error_code) => H3Event::Reset { error_code },
                quiche::h3::Event::GoAway => H3Event::GoAway { id: stream_id },
                quiche::h3::Event::PriorityUpdate => H3Event::PriorityUpdate,
            };
            Ok((stream_id, event))
        }
        Err(quiche::h3::Error::Done) => Err("done".to_string()),
        Err(e) => Err(format!("H3 poll error: {}", h3_error_to_string(e))),
    }
}

#[rustler::nif]
pub fn h3_send_request(
    conn: ResourceArc<ConnectionResource>,
    h3_conn: ResourceArc<H3ConnectionResource>,
    headers: Vec<(Binary, Binary)>,
    fin: bool,
) -> Result<u64, String> {
    let mut quic_conn = conn
        .inner
        .lock()
        .map_err(|e| format!("Lock error: {}", e))?;
    let mut h3_conn = h3_conn
        .inner
        .lock()
        .map_err(|e| format!("Lock error: {}", e))?;

    let headers = build_headers(headers);
    h3_conn
        .send_request(&mut *quic_conn, &headers, fin)
        .map_err(|e| format!("H3 send_request error: {}", h3_error_to_string(e)))
}

#[rustler::nif]
pub fn h3_send_response(
    conn: ResourceArc<ConnectionResource>,
    h3_conn: ResourceArc<H3ConnectionResource>,
    stream_id: u64,
    headers: Vec<(Binary, Binary)>,
    fin: bool,
) -> Result<(), String> {
    let mut quic_conn = conn
        .inner
        .lock()
        .map_err(|e| format!("Lock error: {}", e))?;
    let mut h3_conn = h3_conn
        .inner
        .lock()
        .map_err(|e| format!("Lock error: {}", e))?;

    let headers = build_headers(headers);
    h3_conn
        .send_response(&mut *quic_conn, stream_id, &headers, fin)
        .map_err(|e| format!("H3 send_response error: {}", h3_error_to_string(e)))
}

#[rustler::nif]
pub fn h3_send_body(
    conn: ResourceArc<ConnectionResource>,
    h3_conn: ResourceArc<H3ConnectionResource>,
    stream_id: u64,
    body: Binary,
    fin: bool,
) -> Result<(), String> {
    let mut quic_conn = conn
        .inner
        .lock()
        .map_err(|e| format!("Lock error: {}", e))?;
    let mut h3_conn = h3_conn
        .inner
        .lock()
        .map_err(|e| format!("Lock error: {}", e))?;

    h3_conn
        .send_body(&mut *quic_conn, stream_id, body.as_slice(), fin)
        .map(|_| ())
        .map_err(|e| format!("H3 send_body error: {}", h3_error_to_string(e)))
}

#[rustler::nif]
pub fn h3_recv_body<'a>(
    env: Env<'a>,
    conn: ResourceArc<ConnectionResource>,
    h3_conn: ResourceArc<H3ConnectionResource>,
    stream_id: u64,
    max_bytes: usize,
) -> Result<Binary<'a>, String> {
    let mut quic_conn = conn
        .inner
        .lock()
        .map_err(|e| format!("Lock error: {}", e))?;
    let mut h3_conn = h3_conn
        .inner
        .lock()
        .map_err(|e| format!("Lock error: {}", e))?;

    let mut binary =
        OwnedBinary::new(max_bytes).ok_or_else(|| "Failed to allocate binary".to_string())?;
    match h3_conn.recv_body(&mut *quic_conn, stream_id, binary.as_mut_slice()) {
        Ok(len) => {
            if len < max_bytes && !binary.realloc(len) {
                return Err("Failed to shrink binary".to_string());
            }
            Ok(Binary::from_owned(binary, env))
        }
        Err(quiche::h3::Error::Done) => Err("done".to_string()),
        Err(e) => Err(format!("H3 recv_body error: {}", h3_error_to_string(e))),
    }
}

#[rustler::nif]
pub fn h3_send_goaway(
    conn: ResourceArc<ConnectionResource>,
    h3_conn: ResourceArc<H3ConnectionResource>,
    stream_id: u64,
) -> Result<(), String> {
    let mut quic_conn = conn
        .inner
        .lock()
        .map_err(|e| format!("Lock error: {}", e))?;
    let mut h3_conn = h3_conn
        .inner
        .lock()
        .map_err(|e| format!("Lock error: {}", e))?;

    h3_conn
        .send_goaway(&mut *quic_conn, stream_id)
        .map_err(|e| format!("H3 send_goaway error: {}", h3_error_to_string(e)))
}

#[rustler::nif]
pub fn h3_extended_connect_enabled_by_peer(
    h3_conn: ResourceArc<H3ConnectionResource>,
) -> Result<bool, String> {
    let h3_conn = h3_conn
        .inner
        .lock()
        .map_err(|e| format!("Lock error: {}", e))?;
    Ok(h3_conn.extended_connect_enabled_by_peer())
}

#[rustler::nif]
pub fn h3_dgram_enabled_by_peer(
    conn: ResourceArc<ConnectionResource>,
    h3_conn: ResourceArc<H3ConnectionResource>,
) -> Result<bool, String> {
    let quic_conn = conn
        .inner
        .lock()
        .map_err(|e| format!("Lock error: {}", e))?;
    let h3_conn = h3_conn
        .inner
        .lock()
        .map_err(|e| format!("Lock error: {}", e))?;
    Ok(h3_conn.dgram_enabled_by_peer(&*quic_conn))
}
